### PR TITLE
Add double map API for OSX

### PR DESCRIPTION
**Pull request by bragaigor**
Update double map API to allow to reserve contiguous address
at a FIXED address.

Add release double map region API to free memory associated
to double map accordingly

Include extra test to test new version of double map API.

Refactor double map API to generalize methods and variables names.
Discontiguous arraylets from balanced GC policy is one use case
of double mapping API, but the API can be used for any consumer
that wants to double map multiple regions of memory into a single
contiguous region.

`mmap(2)` in OSX is limitted where it won't allow a mapping that's
backed by a file (have a file descriptor) and huge pages to
work together. That's because huge pages flag is specified
using the file descriptor field; therefore, we cannot have
double mapping, which requires the heap to be backed by a
file descriptor, and huge pages since both use the same field.
In such scenario, if huge pages is available and double map
is requested, huge pages takes precedence over double mapping.

Signed-off-by: Igor Braga <higorb1@gmail.com>

### CODE REVIEW

**Comment by bragaigor** [comment]
cc @amicic @dmitripivkine

**Comment by bragaigor** [comment]
@genie-omr build all

File: port/common/omrport.tdf
Comment (amicic) [inline port/common/omrport.tdf:1593]: these trace point should match methods name more closely

File: gc/base/Heap.hpp
Comment (amicic) [inline gc/base/Heap.hpp:120]: minor (spelling): preferRedAddress (here and many other spots)

File: port/linux/omrvmem.c
Comment (amicic) [inline port/linux/omrvmem.c:1315]: now that you are changing signature, you have a chance to remove byteAmount. you don't have to. it's just a personal preference

**Comment by amicic** [comment]
adding @pshipton (or to delegate someone else) for another pair of eyes for PortLib part of changes

File: port/linux/omrvmem.c
Comment (bragaigor) [reply port/linux/omrvmem.c:1315]: Actually I'll fix this. It would probably be better to allow the caller to set the size required to be double mapped, and here we can just Assert that it is indeed what we expect:
```
Assert_PRT_true((regionsCount * regionSize) == byteAmount);
```
What do you think?

File: port/linux/omrvmem.c
Comment (pshipton) [inline port/linux/omrvmem.c:1361]: This isn't decremented on failure when the memory is unmapped.

File: port/linux/omrvmem.c
Comment (pshipton) [inline port/linux/omrvmem.c:1352]: Seems it could be useful to retain the OS errno. How about creating new portable error codes instead of using OPFAILED for every failure.

File: port/osx/omrvmem.c
Comment (pshipton) [inline port/osx/omrvmem.c:479]: Seems the last 12 would be more unique than the first 12 digits.

Comment (bragaigor) [reply port/osx/omrvmem.c:479]: I totally agree but there’s not an easy way to truncate from the right side. One option I found was:
```
sprintf(omrtime_current_time_str, "%zu", (uintptr_t)omrtime_current_time_millis(portLibrary) % (uintptr_t)pow(10,12));
sprintf(pthread_self_str, "%zu", (uintptr_t)pthread_self() % (uintptr_t)pow(10,9));
sprintf(filename, "omrvmem_%.12s_%.9s", omrtime_current_time_str, pthread_self_str);
```
But I didn’t want to add extra arithmetics (in this case it would be 4) to just name a variable. Maybe I’m overthinking and it’s okay?

pow(10,12) is actually a constant so it could become  `(uintptr_t)omrtime_current_time_millis(portLibrary) % 1000000000000`

Comment (pshipton) [reply port/osx/omrvmem.c:479]: Seems fine using constant values.

**Review by amicic — APPROVED** [review]

**Comment by bragaigor** [comment]
Kindly asking for final review @fjeremic @vijaysun-omr , as this PR is ready to merge.

**Comment by vijaysun-omr** [comment]
@genie-omr build all

**Comment by bragaigor** [comment]
@genie-omr build all

**Comment by bragaigor** [comment]
@genie-omr build linux_x86-64_cmprssptrs

**Comment by bragaigor** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
Since @pshipton commented on this PR, I would appreciate his approval before considering for merging this.

File: port/common/omrport.tdf
Comment (pshipton) [inline port/common/omrport.tdf:1593]: The group name should not be arraylist but double_map or similar.

File: port/common/omrvmem.c
Comment (pshipton) [inline port/common/omrvmem.c:246]: This comment should not mention arraylet.

File: port/linux/omrvmem.c
Comment (pshipton) [inline port/linux/omrvmem.c:1353]: It's possible to use omrerror_set_last_error() here, and pass in errno as the platformCode. This means the error message is generated from the OS using the platformCode (i.e. `strerror()`), but you can't set a custom message. This can be helpful to avoid having to lookup the errno. In order to identify this location differently from other places that set platformCode, use a specific portableCode. For example `OMRPORT_ERROR_VMEM_CONTIGUOUS_DOUBLE_MAP_FAILED`, 'OMRPORT_ERROR_VMEM_DOUBLE_MAP_ADDRESS_FAILED`, etc.

Comment (bragaigor) [reply port/linux/omrvmem.c]: Sounds good, I have also updated the other two instances below that also use `error_set_last_error_with_message_format()` to use `error_set_last_error()` instead. And I did the same for the osx side.

**Comment by bragaigor** [comment]
With the effort of generalizing the code as well as keeping it consistent with the upcoming changes, I have also introduced a new macro `OMR_GC_DOUBLE_MAP_REGIONS` which will replace `OMR_GC_DOUBLE_MAP_ARRAYLETS`. Does this make sense @pshipton or is this too big of a change?

**Comment by pshipton** [comment]
It's the generic port library that shouldn't refer to arraylets. I think the OMR_GC_DOUBLE_MAP_ARRAYLETS is only used by the GC code, which is fine.

**Review by pshipton — APPROVED** [review]

**Comment by pshipton** [comment]
@vijaysun-omr I have approved it now.

**Comment by bragaigor** [comment]
I just found one comment that made reference to arraylet and already updated that. Just double checked everything, squashed, and it's good to go @vijaysun-omr

**Comment by bragaigor** [comment]
@genie-omr build all

**Comment by vijaysun-omr** [comment]
Thanks for the reviews, all the tests seem to have passed and the travis CI issue seems to have gone on for several hours. I don't want to hold this up any longer for some infra issue and therefore I am merging.

**Pull request merged**
